packages/graph-editor/src/hooks/useCopyPaste.ts

```diff
@@ -0,0 +1,195 @@
+import { v4 as uuidv4 } from 'uuid';
+
+function useCopyPaste() {
+  const graphRef = useSelector(graphEditorSelector);
+  const reactFlowInstance = useReactFlow();
+  const toast = useToast();
+
+  const copySelectedNodes = useCallback(() => {
+    if (!graphRef) return;
+
+    const flow = graphRef.getFlow();
+    const graph = graphRef.getGraph();
+    const selectedNodes = flow.getNodes().filter((node) => node.selected);
+    const selectedNodeIds = new Set(selectedNodes.map((node) => node.id));
+
+    // create a subgraph with only the selected nodes
+    const subGraph = graph.serialize();
+    subGraph.nodes = subGraph.nodes.filter((node) =>
+      selectedNodeIds.has(node.id),
+    );
+    const originalEdges = subGraph.edges;
+    subGraph.edges = originalEdges
+      .filter(
+        (edge) =>
+          selectedNodeIds.has(edge.source) && selectedNodeIds.has(edge.target),
+      )
+      .map((edge) => ({
+        id: edge.id,
+        source: edge.source,
+        target: edge.target,
+        sourceHandle: edge.sourceHandle,
+        targetHandle: edge.targetHandle,
+      }));
+
+    // remove viewport from annotations
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const { 'ui.viewport': _, ...restAnnotations } = subGraph.annotations;
+    subGraph.annotations = restAnnotations;
+
+    const serialized = JSON.stringify(subGraph);
+
+    navigator.clipboard
+      .writeText(serialized)
+      .then(() => {
+        toast({
+          title: 'Copied',
+          description: `${selectedNodes.length} nodes copied to clipboard`,
+        });
+      })
+      .catch((error) => {
+        toast({
+          title: 'Error',
+          description: `Failed to copy: ${(error as Error).message}`,
+        });
+      });
+  }, [graphRef, toast]);
+
+  const pasteFromClipboard = useCallback(() => {
+    if (!graphRef) return;
+
+    navigator.clipboard.readText().then(async (text) => {
+      try {
+        const { nodes: sourceNodes, edges: sourceEdges } = JSON.parse(text);
+
+        if (!Array.isArray(sourceNodes) || !Array.isArray(sourceEdges)) {
+          toast({
+            title: 'Error',
+            description: 'Invalid clipboard data',
+          });
+          return;
+        }
+
+        // get the center of the visible area in flow coordinates
+        const topLeft = reactFlowInstance.screenToFlowPosition({ x: 0, y: 0 });
+        const bottomRight = reactFlowInstance.screenToFlowPosition({
+          x: window.innerWidth,
+          y: window.innerHeight,
+        });
+
+        const viewportCenter = {
+          x: (topLeft.x + bottomRight.x) / 2,
+          y: (topLeft.y + bottomRight.y) / 2,
+        };
+
+        // calculate bounding box of source nodes (these are already in flow coordinates)
+        const pasteBounds = sourceNodes.reduce(
+          (acc, node) => ({
+            minX: Math.min(acc.minX, node.annotations?.['ui.position.x'] || 0),
```

Comment:
We should use the `xpos` and `ypos`, etc annotations so we can reduce the amount of hardcoded strings